subprojects/core/src/main/java/org/gradle/internal/classpath/AccessTrackingProperties.java

```diff
@@ -113,47 +123,160 @@ public void forEach(BiConsumer<? super Object, ? super Object> action) {
     @Override
     public void replaceAll(BiFunction<? super Object, ? super Object, ?> function) {
         reportAggregatingAccess();
-        delegate.replaceAll(function);
+        synchronized (delegate) {
+            delegate.replaceAll((k, v) -> {
+                Object newValue = function.apply(k, v);
+                if (v != newValue) {
```

Comment:
🤔 I wonder if we shouldn't be at least handling primitive types via `equals`

Reply:
It's a good idea. However, TBH, I don't think anyone would use this method to modify system properties. Because of this, I'm inclined to keep "correct and simple, though less effective" implementation. Replacing with "equals" everywhere would be incorrect in some (albeit very corner) cases, and adding a type check to handle primitives wouldn't be simple.

Reply:
I had a much more complex thing in mind, thank you for pointing me to `isPrimitive`, it simplifies everything. I'll use it then.

Reply:
`isPrimitive` wasn't exactly a match (as it is `true` for, e.g. `int.class` and we have boxed `Integer` here), but, fortunately, there is a helper.